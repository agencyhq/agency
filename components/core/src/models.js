import knexfile from '../knexfile.js'
import Knex from 'knex'
import Bookshelf from 'bookshelf'

const knex = Knex(knexfile[process.env.NODE_ENV] || knexfile.development)
const bookshelf = Bookshelf(knex)

const Executions = bookshelf.model('Executions', {
  tableName: 'executions',
  result () {
    return this.hasOne('Results', 'id', 'id')
  }
})

const Results = bookshelf.model('Results', {
  tableName: 'results'
})

const Rules = bookshelf.model('Rules', {
  tableName: 'rules'
})

const Tokens = bookshelf.model('Tokens', {
  tableName: 'tokens'
})

export default {
  knex,
  bookshelf,
  Executions,
  Results,
  Rules,
  Tokens
}
