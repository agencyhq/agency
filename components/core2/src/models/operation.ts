export type OperationStatus = 'requested' | 'claimed' | 'processing' | 'completed' | 'failed';

export class Operation {
  id: string;
  action: string;
  params: Record<string, unknown>;
  status: OperationStatus;
  result: Record<string, unknown>;
  createdAt: Date;
  updatedAt: Date;
}
