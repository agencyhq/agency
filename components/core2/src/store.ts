import { Operation } from './models/operation';

export class Storage {
  private operations: Operation[] = [];

  loadOperations(operations: Operation[]): void {
    this.operations = [...operations];
  }

  getOperations(): Operation[] {
    return this.operations;
  }

  getOperation(id: string): Operation | undefined {
    return this.operations.find((operation) => operation.id === id);
  }

  createOperation(operation: Operation): Operation {
    this.operations.push(operation);
    return operation;
  }

  updateOperation(operation: Operation): Operation {
    const index = this.operations.findIndex((op) => op.id === operation.id);
    this.operations[index] = operation;
    return operation;
  }

  deleteOperation(id: string): Operation | undefined {
    const index = this.operations.findIndex((op) => op.id === id);
    if (index === -1) {
      return undefined;
    }
    return this.operations.splice(index, 1)[0];
  }

  clearOperations(): void {
    this.operations = [];
  }
}
