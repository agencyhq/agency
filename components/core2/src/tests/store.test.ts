import { describe, expect, it } from '@jest/globals';

import { Operation } from '../models/operation';
import { Storage } from '../store';
import fixtures from './__fixtures__/operations';

describe('store', () => {
  let storage: Storage;

  it('should return operations', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);
    expect(storage.getOperations()).toMatchSnapshot();
  });

  it('should return operation', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);
    expect(storage.getOperation('1')).toMatchSnapshot();
  });

  it('should create operation', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);

    const operation: Operation = {
      id: '5',
      action: 'create',
      params: { name: 'John Doe' },
      status: 'completed',
      result: { id: '5', name: 'John Doe' },
      createdAt: new Date('2021-01-01'),
      updatedAt: new Date('2021-01-01'),
    };

    expect(storage.createOperation(operation)).toMatchSnapshot();
    expect(storage.getOperations()).toMatchSnapshot();
  });

  it('should update operation', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);

    const operation: Operation = {
      id: '3',
      action: 'create',
      params: { name: 'John Doe' },
      status: 'completed',
      result: { id: '3', name: 'John Doe' },
      createdAt: new Date('2021-01-01'),
      updatedAt: new Date('2021-01-01'),
    };

    expect(storage.updateOperation(operation)).toMatchSnapshot();
    expect(storage.getOperations()).toMatchSnapshot();
  });

  it('should delete operation', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);

    expect(storage.deleteOperation('1')).toMatchSnapshot();
    expect(storage.getOperations()).toMatchSnapshot();
  });

  it('should return undefined if deleting operation that does not exist', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);

    expect(storage.deleteOperation('5')).toMatchSnapshot();
    expect(storage.getOperations()).toMatchSnapshot();
  });

  it('should clear operations', () => {
    storage = new Storage();
    storage.loadOperations(fixtures);

    storage.clearOperations();
    expect(storage.getOperations()).toMatchSnapshot();
  });
});
