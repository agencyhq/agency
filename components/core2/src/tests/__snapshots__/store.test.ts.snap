// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should clear operations 1`] = `[]`;

exports[`store should create operation 1`] = `
{
  "action": "create",
  "createdAt": 2021-01-01T00:00:00.000Z,
  "id": "5",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "5",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": 2021-01-01T00:00:00.000Z,
}
`;

exports[`store should create operation 2`] = `
[
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "1",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "1",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "update",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "2",
    "params": {
      "id": "1",
      "name": "Jane Doe",
    },
    "result": {
      "id": "1",
      "name": "Jane Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "delete",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "3",
    "params": {
      "id": "1",
    },
    "result": {
      "id": "1",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "5",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "5",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
]
`;

exports[`store should delete operation 1`] = `
{
  "action": "create",
  "createdAt": 2021-01-01T00:00:00.000Z,
  "id": "1",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "1",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": 2021-01-01T00:00:00.000Z,
}
`;

exports[`store should delete operation 2`] = `
[
  {
    "action": "update",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "2",
    "params": {
      "id": "1",
      "name": "Jane Doe",
    },
    "result": {
      "id": "1",
      "name": "Jane Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "delete",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "3",
    "params": {
      "id": "1",
    },
    "result": {
      "id": "1",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
]
`;

exports[`store should return operation 1`] = `
{
  "action": "create",
  "createdAt": 2021-01-01T00:00:00.000Z,
  "id": "1",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "1",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": 2021-01-01T00:00:00.000Z,
}
`;

exports[`store should return operations 1`] = `
[
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "1",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "1",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "update",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "2",
    "params": {
      "id": "1",
      "name": "Jane Doe",
    },
    "result": {
      "id": "1",
      "name": "Jane Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "delete",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "3",
    "params": {
      "id": "1",
    },
    "result": {
      "id": "1",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
]
`;

exports[`store should return undefined if deleting operation that does not exist 1`] = `undefined`;

exports[`store should return undefined if deleting operation that does not exist 2`] = `
[
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "1",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "1",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "update",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "2",
    "params": {
      "id": "1",
      "name": "Jane Doe",
    },
    "result": {
      "id": "1",
      "name": "Jane Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "delete",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "3",
    "params": {
      "id": "1",
    },
    "result": {
      "id": "1",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
]
`;

exports[`store should update operation 1`] = `
{
  "action": "create",
  "createdAt": 2021-01-01T00:00:00.000Z,
  "id": "3",
  "params": {
    "name": "John Doe",
  },
  "result": {
    "id": "3",
    "name": "John Doe",
  },
  "status": "completed",
  "updatedAt": 2021-01-01T00:00:00.000Z,
}
`;

exports[`store should update operation 2`] = `
[
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "1",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "1",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "update",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "2",
    "params": {
      "id": "1",
      "name": "Jane Doe",
    },
    "result": {
      "id": "1",
      "name": "Jane Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
  {
    "action": "create",
    "createdAt": 2021-01-01T00:00:00.000Z,
    "id": "3",
    "params": {
      "name": "John Doe",
    },
    "result": {
      "id": "3",
      "name": "John Doe",
    },
    "status": "completed",
    "updatedAt": 2021-01-01T00:00:00.000Z,
  },
]
`;
